File: www/linked_list_wasm.js
// Mock WebAssembly module for linked list
function LinkedListModule() {
    // Create a simple linked list implementation in JavaScript
    let elements = [];
    const maxSize = 20;

    return Promise.resolve({
        isEmpty: function() {
            return elements.length === 0;
        },
        isFull: function() {
            return elements.length >= maxSize;
        },
        size: function() {
            return elements.length;
        },
        maxSize: function() {
            return maxSize;
        },
        insertAtBeginning: function(value) {
            if (elements.length >= maxSize) {
                return false;
            }
            elements.unshift(value);
            return true;
        },
        insertAtEnd: function(value) {
            if (elements.length >= maxSize) {
                return false;
            }
            elements.push(value);
            return true;
        },
        insertAtPosition: function(value, position) {
            if (elements.length >= maxSize || position > elements.length) {
                return false;
            }
            elements.splice(position, 0, value);
            return true;
        },
        deleteAtBeginning: function() {
            if (elements.length === 0) {
                return false;
            }
            elements.shift();
            return true;
        },
        deleteAtEnd: function() {
            if (elements.length === 0) {
                return false;
            }
            elements.pop();
            return true;
        },
        deleteAtPosition: function(position) {
            if (elements.length === 0 || position >= elements.length) {
                return false;
            }
            elements.splice(position, 1);
            return true;
        },
        search: function(value) {
            return elements.indexOf(value);
        },
        getElements: function() {
            return elements;
        },
        clear: function() {
            elements = [];
        }
    });
} 

File: src/linked_list_wasm.cpp
#include <emscripten/bind.h>
#include <emscripten/emscripten.h>
#include "../include/LinkedList.h"
#include <vector>
#include <string>

LinkedList linkedList(20);

bool isEmpty() {
    return linkedList.isEmpty();
}

bool isFull() {
    return linkedList.isFull();
}

size_t size() {
    return linkedList.getSize();
}

size_t maxSize() {
    return linkedList.getMaxSize();
}

bool insertAtBeginning(int value) {
    return linkedList.insertAtBeginning(value);
}

bool insertAtEnd(int value) {
    return linkedList.insertAtEnd(value);
}

bool insertAtPosition(int value, size_t position) {
    return linkedList.insertAtPosition(value, position);
}

bool deleteAtBeginning() {
    return linkedList.deleteAtBeginning();
}

bool deleteAtEnd() {
    return linkedList.deleteAtEnd();
}

bool deleteAtPosition(size_t position) {
    return linkedList.deleteAtPosition(position);
}

int search(int value) {
    return linkedList.search(value);
}

emscripten::val getElements() {
    std::vector<int> elements = linkedList.getElements();
    return emscripten::val(emscripten::typed_memory_view(
        elements.size(),
        elements.data()
    ));
}

void clear() {
    linkedList.clear();
}

EMSCRIPTEN_BINDINGS(linked_list_module) {
    emscripten::function("isEmpty", &isEmpty);
    emscripten::function("isFull", &isFull);
    emscripten::function("size", &size);
    emscripten::function("maxSize", &maxSize);
    emscripten::function("insertAtBeginning", &insertAtBeginning);
    emscripten::function("insertAtEnd", &insertAtEnd);
    emscripten::function("insertAtPosition", &insertAtPosition);
    emscripten::function("deleteAtBeginning", &deleteAtBeginning);
    emscripten::function("deleteAtEnd", &deleteAtEnd);
    emscripten::function("deleteAtPosition", &deleteAtPosition);
    emscripten::function("search", &search);
    emscripten::function("getElements", &getElements);
    emscripten::function("clear", &clear);
}
